WIP: Tanzania dataset integration

delhomer opened the pull request:
This PR introduces the Tanzania dataset into deeposlandia framework.

The main modifications arise as follows:
- a new module has been defined for defining Tanzania buildings (`deeposlandia/buildings.py`). It inherits from the `Dataset` class. The main part of the job is done here, as this dataset involves geo-referenced images. The project then needs to consider geographic tools like GDAL and geopandas.
- some minor modifications were needed to update docstrings and if-case in other modules (`datagen.py`, `generator.py`, `train.py`, `paramoptim.py`, `inference.py`).
- the web application has been updated (new webpage, reference on the index.html page).
- the README.md has been enriched with the new dataset description. Additionally, a new markdown document describes how to add new datasets, by listing all required operations.
- some new tests that consider Tanzania dataset have been designed, and a corresponding data sample has been added to `tests/data` folder.

The PR is still on processing, as it misses : 
- the training of a model (the README.md will have to be updated with a more convincing prediction example...)

delhomer committed:
[dataset] Add a module for considering a new dataset

The dataset relies on aerial image from Tanzania. One has a bunch of
high-resolution images and three labels to discriminate: complete
buildings, incomplete buildings and foundations (plus a
background). Further details may be get at
https://blog.werobotics.org/2018/08/06/welcome-to-the-open-ai-tanzania-challenge/

delhomer committed:
[dataset] Introduce Tanzania dataset into datagen.py program

delhomer committed:
[datagen/train/inference] Introduce tanzania (resp. open_ai_tanzania) as a new possible dataset

delhomer committed:
[dataset] Build labelled image from a numpy mask instead of a PIL.Image

delhomer committed:
[dataset] Create a _preprocess_tile method so as to decompose preprocessing

delhomer committed:
[dataset] update config sample with tanzania dataset link

delhomer committed:
[webapp] Add Tanzania dataset to the Flask application

delhomer committed:
[readme] update readme with a mention to Tanzania dataset

delhomer committed:
[dataset] fix tanzania background color

delhomer committed:
[dataset] reorganize dataset modules

delhomer committed:
[test] add some tests for managing Tanzania dataset

delhomer committed:
[test] add tanzania sampled files for testing purpose

delhomer committed:
[dataset] update the file  with test-related instructions :-)

delhomer commented:
@garaud I'm working on training a decent semantic segmentation model, before closing the PR. However, could it be possible to request your opinion on the remaining aspects of this PR?

garaud commented on deeposlandia/datasets/add_a_dataset.md:
```diff
@@ -0,0 +1,80 @@
+
+Following folders must be created and maintained:
+
++ `data/dataset/input/training/images` must contains raw training images
```
must contain, without the "s"

delhomer replied:
Fixed, as all the following lines that contain the same mistake.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+
+from multiprocessing import Pool
+import numpy as np
+import os
```
First the import of the stdlib, i.e. multiprocessing and os.

Then a new line and the other dependencies such as PIL and numpy.

delhomer replied:
Fixed in all the modules.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+
+class AerialDataset(dataset.Dataset):
+    """Dataset structure inspired from AerialImageDataset, a dataset released
+    by INRIA
```
INRIA => Inria.

delhomer replied:
Fixed.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+    tile_size : int
+        Size of the tiles into which each raw images is decomposed during
+    dataset population (height=width)
+
```
remaining useless new line in the docstring

delhomer replied:
Done.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+        """ Class constructor ; instanciates a AerialDataset as a standard
+        Dataset which is completed by a glossary file that describes the
+        dataset labels and images
+
```
useless newline in the docstring

delhomer replied:
Fixed.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+from deeposlandia.datasets import dataset
+from deeposlandia import utils
+
+
```
The logger should be instantiated here. You don't have to use the one which is in the `utils` module.

delhomer replied:
I do not understand this point. Do you suggest to duplicate the logger object as many times as the number of module?

I would have guessed that it is better to have one logger for the complete app.

garaud replied:
From https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial

> A good convention to use when naming loggers is to use a module-level logger, in each module which uses logging, named as follows:

```python
logger = logging.getLogger(__name__)
```

garaud replied:
one instance of `logger` by module

delhomer replied:
Ok thanks for the reference, I will read it and propose code updates!

delhomer replied:
Done in a new commit (would have been better in another PR, anyway it is done!).

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+                                         for x in image_list_longname])
+        self.image_info = [item for sublist in self.image_info
+                           for item in sublist]
+        utils.logger.info(("Saved {} images in the preprocessed dataset."
```
see my comment about the logger

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+        image_list_longname = [os.path.join(input_dir, "images", l)
+                               for l in image_list if not l.startswith('.')][:nb_images]
+        utils.logger.info(("Getting {} images to preprocess..."
+                           "").format(len(image_list_longname)))
```
The logger should use the "%s" format string, not the `format` method.

e.g. `logger.info("Save the file '%s'", filename)`

delhomer replied:
Have you got a source, or a pep, that advises not to use the `format` method?

delhomer replied:
And what about the f-string method? (see [PEP498](https://www.python.org/dev/peps/pep-0498/#how-to-denote-f-strings))

delhomer replied:
I've changed all the logger formatted strings from `.format()` method to f-strings.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+        aggregate : bool
+            Label aggregation parameter, useless for this dataset, but kept for
+        class method genericity
+        labelling: boolean
```
`labelling : boolean` missing a whitespace

delhomer replied:
Fixed.

garaud commented on deeposlandia/datasets/aerial.py:
```diff
@@ -0,0 +1,144 @@
+            label_filename = image_filename.replace("images/", "gt/")
+            img_out = Image.open(label_filename) 
+            buffer_tiles = []
+            for x in range(0, raw_img_size, self.tile_size):
```
this double loop looks like with the previous one. Don't you think it's possible to refactor this ?

delhomer replied:
It certainly is. Could we improve that particular point in a subsequent PR?

delhomer replied:
I have just created a new issue to address this problem. I propose to close the discussion in this PR scope.

garaud commented on deeposlandia/datasets/dataset.py:
```diff
@@ -0,0 +1,161 @@
+"""Encapsulates datasets in a generic Dataset class, and in some more specific
+classes that inherit from it
+
```
useless new line

garaud replied:
Maybe this mother class should be in the `__init__.py` module of the `dataset` sub-package.

delhomer replied:
New line removed, and class moved to `__init__.py`.

delhomer replied:
Is it a good practice to put so much material in a `__init__.py` module?

garaud commented on deeposlandia/datasets/dataset.py:
```diff
@@ -0,0 +1,161 @@
+import numpy as np
+
+from deeposlandia import utils
+
```
pep8 : 2 blank lines between the imports and the class

delhomer replied:
Fixed.

garaud commented on deeposlandia/datasets/dataset.py:
```diff
@@ -0,0 +1,161 @@
+    image_size : int
+        Size of considered images (height=width), raw images will be resized during the
+    preprocessing
+
```
useless newline

delhomer replied:
Fixed.

garaud commented on deeposlandia/datasets/dataset.py:
```diff
@@ -0,0 +1,161 @@
+
+from deeposlandia import utils
+
+class Dataset:
```
If the method `populate` must be implemented, the code should raise an error if it's not the case. Maybe we should use the `abstractmethod`  decorator to force this.

This should be the case for all methods or properties we must have in order to have a well-defined subclass.

delhomer replied:
I created one single abstract method, namely `populate`, by using `@abc.abstractmethod` decorator.

garaud commented on deeposlandia/datasets/mapillary.py:
```diff
@@ -0,0 +1,166 @@
+            if aggregate:
+                img_out = self.group_image_label(img_out)
+
+            labels = utils.label_building(img_out,
```
`label_building` is confusing with the notion of "building" in some other datasets. If it's the verb and not the noun, I prefer to use the infinitive form "build".

delhomer replied:
Fixed, and renamed as `build_labels`.

garaud commented on deeposlandia/generator.py:
```diff
@@ -152,7 +153,7 @@ def create_generator(dataset, model, datapath, image_size, batch_size, label_con
 
     """
-    if not dataset in ['shapes', 'mapillary', 'aerial']:
+    if not dataset in ['shapes', 'mapillary', 'aerial', 'tanzania', 'open_ai_tanzania']:
```
maybe this list should be a global variable

delhomer replied:
You're right, I put it into `deeposlandia/datasets/__init__.py` module. Really useful when associated with the argument parser too!

garaud commented on deeposlandia/inference.py:
```diff
@@ -48,7 +48,8 @@ def add_program_arguments(parser):
     parser.add_argument('-D', '--dataset',
                         required=True,
-                        help="Dataset type (either mapillary, shapes or aerial)")
+                        help=("Dataset type (either mapillary, shapes, aerial"
```
you pass a specific list of choices with argparse. See https://docs.python.org/3/library/argparse.html#choices

delhomer replied:
Modified. Thanks for this suggestion!

garaud commented on deeposlandia/train.py:
```diff
@@ -35,7 +35,8 @@ def add_instance_arguments(parser):
                         required=True,
-                        help="Dataset type (either mapillary, shapes or aerial)")
+                        help=("Dataset type (either mapillary, shapes, aerial"
+                              " or tanzania)"))
```
I think we can group some option variables in a same place instead of having 3 different `add_argument` in 3 different modules. The CLI must be centralized in a dedicated module. We can open a new issue for that.

delhomer replied:
I agree to report this to a further issue/PR. Could you write it please?

garaud replied:
done https://github.com/Oslandia/deeposlandia/issues/90

delhomer replied:
Thank you for creating the issue!

garaud commented on deeposlandia/utils.py:
```diff
@@ -62,6 +62,29 @@ def label_building(filtered_image, label_ids, dataset='mapillary'):
+    dict
+        Label ids which occur (or not) in the image
+    """
+    tile_labels = [s.lower() for s in items.unique().tolist()]
```
maybe you can do this `items.unique().str.lower().tolist()`

delhomer replied:
The `geojson_label_building` is even useless, I removed it. This particular point is then solved.

delhomer committed:
reorganize module imports

delhomer committed:
[dataset] Move generic Dataset class to local __init__.py module

delhomer committed:
[dataset] Store available datasets in a dedicated variable

delhomer committed:
[dataset] change a method name from label_building to build_labels and delete the useless method geojson_label_building

delhomer committed:
[dataset] various docstring fixes

delhomer committed:
Refactor the logging process

Generalize the usage of daiquiri (which was used only in webapp before
this commit), and declare a logger in each module.
